### DISCUSSION TITLE
Naming convention for CB-TB resource object (acceptable characters)

### DISCUSSION THREAD

**Discussion by seokho-son** [cloud-barista/cb-tumblebug#554] [General]
### 개요
CB-TB는 다양한 Cloud service provider(CSP) 와 연동하므로, 
클라우드 리소스 생성시 각 CSP의 리소스 네이밍 컨벤션을 준수할 필요가 있음. 
이에 따라 CB-TB의 리소스 오브젝트의 네이밍 컨벤션을 지정하여, 문제 발생 소지를 최소화하고자 함.
(참고: CB-TB는 사용자가 지정한 `name` 값을 활용하여 `id` 값을 생성함)

### CB-TB 오브젝트 일반적 규칙

- 정규 표현식 : `[a-z]([-a-z0-9]*[a-z0-9])?`
- 허용 문자
  - 영어 소문자
  - 숫자(시작 문자로는 사용할 수 없음)
  - 하이픈(시작/끝 문자로는 사용할 수 없음)
- 허용하지 않는 문자
  - 허용 이외의 문자 (공백, 언더스코어, 마침표, 콜론, 슬래시 등)

### CB-TB 오브젝트별 상세 규칙

- SecurityGroup 오브젝트: `-deli-` 는 사용 불가 (CB-Spider에서 특수문자로 사용하고 있음)

## CB-TB 오브젝트의 `name` 필드 문자열 최대 길이

- 논의 및 적용 필요
- GCP 등은 63자 제한이 있으므로, 63자 이하로 지정이 필요함

## 참고자료
- GCP Naming resources (https://cloud.google.com/compute/docs/naming-resources)
> For resources on Compute Engine, the resource name must meet the following format (unless specified differently in the documentation for creating that resource):
> - Be 1-63 characters in length.
> - Comply with RFC 1035 conventions.
> - Match the regular expression `[a-z]([-a-z0-9]*[a-z0-9])?`. This means that the first character must be a lowercase letter, and all the following characters must be hyphens, lowercase letters, or digits, except the last character, which cannot be a hyphen.

- AWS Tag naming limits and requirements (https://docs.aws.amazon.com/general/latest/gr/aws_tagging.html)
> The following basic naming and usage requirements apply to tags:
> - The tag value must be a minimum of 0 and a maximum of 256 Unicode characters in UTF-8. (Some services don't permit tags with an empty value (length of 0))
> - Allowed characters can vary by AWS service. For information about what characters you can use to tag resources in a particular AWS service, see its documentation. In general, allowed characters in tags are letters, numbers, spaces representable in UTF-8, and the following characters: `_ . : / = + - @ `.
> - Tag keys and values are case sensitive. As a best practice, decide on a strategy for capitalizing tags, and consistently implement that strategy across all resource types. For example, decide whether to use Costcenter, costcenter, or CostCenter, and use the same convention for all tags. Avoid using similar tags with inconsistent case treatment.

- Alibaba Names of instances (https://www.alibabacloud.com/help/faq-detail/38580.htm)

> - The name can contain only letters, digits, and hyphens (-).
> - The name is case-insensitive.
> - The name must start with a letter and cannot end with a hyphen (-).
> - The name must be 3 to 16 bytes in length.

- Azure (https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules#microsoftcompute)
> - When retrieving resource names using various APIs, returned values may display different case values than what is listed in the valid characters table.
> virtualMachines : 
> - Can't use space or these characters: `~ ! @ # $ % ^ & * ( ) = + _ [ ] { } \ | ; : . ' " , < > / ?`
> - Length: 1-15 (Windows), 1-64 (Linux)

## 관련 ISSUE 및 PR
- https://github.com/cloud-barista/cb-tumblebug/issues/445
- https://github.com/cloud-barista/cb-tumblebug/pull/527

**Comment by seokho-son** [comment]
@powerkimhub 
CB-Spider에서 ID 또는 Name을 임의로 생성하여 CSP에 전달하는 항목(ex: SG)  공유를 부탁드립니다. (생성 방식)
CSP에게 전달되는 string의 길이에 대해서 대략 파악하기 위함 입니다.

**Comment by powerkimhub** [comment]
@seokho-son 

- 재확인해보니 자원의 Name 관련 처리가 2가지네요. 
- 현황은 아래와 같습니다. 

#### [SG Name]
- 용도: SG 관리가 VPC에 종속적인 CSP와 독립적으로 운영하는 CSP가 공존
  - 의존성을 표시하기 위한 방법으로 SG Name에 소유하는 VPC 이름을 함께 태깅함. 
- 현황: SG Name 가능 길이가 속하는 VPC Name의 길이에 따라 유동적임.
- 관련 constant: SG_DELIMITER
  
#### [Subnet Name]
- 용도: VPC와 Subnet 관계성 유지(SG와 유사하게 관계성이 없는 CSP가 있었음)
- subnet 이름 입력시 앞부분 제약 단어 존재 
  - 사용자가 지정한 subnet 이름 앞부분에 `subnet:` 을 포함하면 튕겨냄.
- 실제 자원 이름은 추가되지 않고, Spider Server에서만 활용되고 있음.
- 현황: subnet 이름 길이에는 영향을 주지 않고, subnet 이름 부여시 제약 존재.
- 관련 constant: SUBNET_PREFIX
 
#### [계획]
- SG: delimiter를 사용하지 않는 방법으로 변경 
- Subnet: prefix를 사용하지 않는 방법으로 변경 
- 시작시점: 4차 오픈컨퍼런스 이후

**Comment by seokho-son** [comment]
내부적으로 Prefix 등을 붙이는 것이 현재는
SG (Prefix: VPC-ID + "-deli-") 와 Subnet (Prefix: "subnet-") 가 있습니다.

M ID 중복 이슈를 해결하기 위해서
CSP 기준 VM ID를
NS-ID + "-" + MCIS-ID + "-" VM-ID 형태로 처리할 예정 입니다.

CSP가 수용할 수 있는 스트링 길이가 60 자 정도라고 가정하면
VM ID에 다소 제약이 더 생길 것 같습니다. (각 오브젝트의 ID가 20자라고 가정하면, 최대 20+20+20+2 = 62)
따라서, 현재 시점에서는 조금 더 방어적으로 잡아주시는 것이 좋을 것 같습니다. (향후 조정 가능성 높음)

15자 정도로 제한하는 것으로 기본 안을 잡았습니다.

**Comment by seokho-son** [comment]
FYI. CloudIt VM ID 글자수 제한 (45)

CC @powerkimhub @jihoon-seo

**Comment by jihoon-seo** [comment]
NCP serverName 은 3~30자 이어야 함


> [CLOUD-BARISTA].[ERROR]: 2021-06-22 14:53:22 CommonNcpFunc.go:65, github.com/cloud-barista/cb-spider/cloud-control-manager/cloud-driver/drivers/ncp/resources.LoggingError() - Status: 400 Bad Request, Body: {"responseError": {
  "returnCode": "905",
  "returnMessage": "Please check your input value : [tb01-ncp-korea1-jhseo-ncp-korea1-jhseo-2]. Length constraints: Minimum length of 3. Maximum length of 30. location : serverName"
}}

**Reply by seokho-son** [reply]
3~ 은 또 첨보는군요 ㅎㅎ
